frontend/pages/upload.vue

```diff
@@ -0,0 +1,407 @@
+    if (!respItem) {
+      throw new Error(`Could not find start item for file ${file.name} - something is probably wrong in the API bizlogic`)
+    }
+    fileStates.value[index].incompleteUploadId = respItem.incomplete_upload_id
+
+    await $axios({
+      method: 'PUT',
+      url: respItem.upload_url,
+      data: file,
+      headers: {
+        'Content-Type': file.type,
+        'x-ms-blob-type': 'BlockBlob',
+      },
+    }).then(() => {
+      fileStates.value[index].status = FileStatus.Uploaded
+    }).catch((e) => {
+      console.log('error uploading file', e)
+      fileStates.value[index].status = FileStatus.Error
+      fileStates.value[index].errorMessage = serializeError(e)
+      errorCode.value = 'Upload Failed'
+      errorMessage.value = 'One or more files could not be uploaded - please delete/resolve them and try again.'
+    })
+  })
+  await Promise.all(uploads)
+
+  let hadError = false
+  for (const fileState of fileStates.value) {
+    if (fileState.status === FileStatus.Uploaded) {
+      continue
+    } else if (fileState.status === FileStatus.Error) {
+      hadError = true
+    } else {
+      throw new Error(`Unexpected file state ${fileState.status}`)
+    }
+  }
+  if (!hadError) {
+    await doParsing()
+  }
+  isProcessing.value = false
+}
+
+const doParsing = async () => {
+  fileStates.value.forEach((_, i) => {
+    fileStates.value[i].status = FileStatus.Validating
+  })
+  await pactaClient.completePortfolioUpload({
+    items: fileStates.value.map((fileState) => ({
+      incomplete_upload_id: presentOrFileBug(fileState.incompleteUploadId),
+    })),
+  })
+  await waitForValidationToCompleteOrTimeout()
+  await cleanUpIncompleteUploads()
+}
+
+const waitForValidationToCompleteOrTimeout = async () => {
+  const timeout = 1000 * 60 * 5
+  const start = Date.now()
+  do {
+    await refreshStateFromIncompleteUploads()
+    if (Date.now() - start > timeout) {
+      throw new Error('Timeout waiting for uploads to complete')
+    }
+    await new Promise((resolve) => setTimeout(resolve, 1000))
+  } while (fileStates.value.some((fileState) => fileState.status === FileStatus.Validating))
+  isProcessing.value = false
+}
+
+const refreshStateFromIncompleteUploads = async () => {
+  const resp = await pactaClient.listIncompleteUploads()
+  console.log('resp', resp)
+  const incompleteUploads = resp.items
+  for (const incompleteUpload of incompleteUploads) {
+    const idx = fileStates.value.findIndex((fileState) => fileState.incompleteUploadId === incompleteUpload.id)
+    // Note - this item might not be in the list if the user hasn't cleaned up prior incomplete portfolios.
+    if (idx >= 0) {
+      if (incompleteUpload.failureCode) {
+        fileStates.value[idx].status = FileStatus.Error
+        fileStates.value[idx].errorMessage = incompleteUpload.failureMessage
+      } else if (incompleteUpload.completedAt) {
+        fileStates.value[idx].status = FileStatus.CleanUp
+      }
+    }
```

Comment:
optional nit: handle error conditions early and unindent main body, e.g.

```suggestion
    // Note - this item might not be in the list if the user hasn't cleaned up prior incomplete portfolios.
  if (idx < 0) {
    continue
  }
  if (incompleteUpload.failureCode) {
    fileStates.value[idx].status = FileStatus.Error
    fileStates.value[idx].errorMessage = incompleteUpload.failureMessage
  } else if (incompleteUpload.completedAt) {
    fileStates.value[idx].status = FileStatus.CleanUp
  }
}
```

Reply:
Great call